From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Sun, 6 Jul 2025 02:38:39 +0300
Subject: [PATCH] C2ME: optimize noise generation

This patch is based on the following mixins:
* "com/ishland/c2me/opts/math/mixin/MixinOctavePerlinNoiseSampler.java"
* "com/ishland/c2me/opts/math/mixin/MixinPerlinNoiseSampler.java"
By: ishland <ishlandmc@yeah.net>
As part of: C2ME (https://github.com/RelativityMC/C2ME-fabric)
Licensed under: MIT (https://opensource.org/licenses/MIT)

diff --git a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
index f30358343a80ce094b79cef5d4443ebd82dc98fa..6bc902e9f15b556a76c023c171ac489ba4939029 100644
--- a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
@@ -11,6 +11,27 @@ public final class ImprovedNoise {
     public final double yo;
     public final double zo;
 
+    // DivineMC start - C2ME: optimize noise generation
+    private static final double[] FLAT_SIMPLEX_GRAD = new double[]{
+        1, 1, 0, 0,
+        -1, 1, 0, 0,
+        1, -1, 0, 0,
+        -1, -1, 0, 0,
+        1, 0, 1, 0,
+        -1, 0, 1, 0,
+        1, 0, -1, 0,
+        -1, 0, -1, 0,
+        0, 1, 1, 0,
+        0, -1, 1, 0,
+        0, 1, -1, 0,
+        0, -1, -1, 0,
+        1, 1, 0, 0,
+        0, -1, 1, 0,
+        -1, 1, 0, 0,
+        0, -1, -1, 0,
+    };
+    // DivineMC end - C2ME: optimize noise generation
+
     public ImprovedNoise(final RandomSource random) {
         this.xo = random.nextDouble() * 256.0;
         this.yo = random.nextDouble() * 256.0;
@@ -38,9 +59,11 @@ public final class ImprovedNoise {
         double x = _x + this.xo;
         double y = _y + this.yo;
         double z = _z + this.zo;
-        int xf = Mth.floor(x);
-        int yf = Mth.floor(y);
-        int zf = Mth.floor(z);
+        // DivineMC start - C2ME: optimize noise generation
+        double xf = Math.floor(x);
+        double yf = Math.floor(y);
+        double zf = Math.floor(z);
+        // DivineMC end - C2ME: optimize noise generation
         double xr = x - xf;
         double yr = y - yf;
         double zr = z - zf;
@@ -53,25 +76,27 @@ public final class ImprovedNoise {
                 fudgeLimit = yr;
             }
 
-            yrFudge = Mth.floor(fudgeLimit / yScale + 1.0E-7F) * yScale;
+            yrFudge = Math.floor(fudgeLimit / yScale + 1.0E-7F) * yScale; // DivineMC - C2ME: optimize noise generation
         } else {
             yrFudge = 0.0;
         }
 
-        return this.sampleAndLerp(xf, yf, zf, xr, yr - yrFudge, zr, yr);
+        return this.sampleAndLerp((int) xf, (int) yf, (int) zf, xr, yr - yrFudge, zr, yr); // DivineMC - C2ME: optimize noise generation
     }
 
     public double noiseWithDerivative(final double _x, final double _y, final double _z, final double[] derivativeOut) {
         double x = _x + this.xo;
         double y = _y + this.yo;
         double z = _z + this.zo;
-        int xf = Mth.floor(x);
-        int yf = Mth.floor(y);
-        int zf = Mth.floor(z);
+        // DivineMC start - C2ME: optimize noise generation
+        double xf = Math.floor(x);
+        double yf = Math.floor(y);
+        double zf = Math.floor(z);
+        // DivineMC end - C2ME: optimize noise generation
         double xr = x - xf;
         double yr = y - yf;
         double zr = z - zf;
-        return this.sampleWithDerivative(xf, yf, zf, xr, yr, zr, derivativeOut);
+        return this.sampleWithDerivative((int) xf, (int) yf, (int) zf, xr, yr, zr, derivativeOut); // DivineMC - C2ME: optimize noise generation
     }
 
     private static double gradDot(final int hash, final double x, final double y, final double z) {
@@ -83,24 +108,69 @@ public final class ImprovedNoise {
     }
 
     private double sampleAndLerp(final int x, final int y, final int z, final double xr, final double yr, final double zr, final double yrOriginal) {
-        int x0 = this.p(x);
-        int x1 = this.p(x + 1);
-        int xy00 = this.p(x0 + y);
-        int xy01 = this.p(x0 + y + 1);
-        int xy10 = this.p(x1 + y);
-        int xy11 = this.p(x1 + y + 1);
-        double d000 = gradDot(this.p(xy00 + z), xr, yr, zr);
-        double d100 = gradDot(this.p(xy10 + z), xr - 1.0, yr, zr);
-        double d010 = gradDot(this.p(xy01 + z), xr, yr - 1.0, zr);
-        double d110 = gradDot(this.p(xy11 + z), xr - 1.0, yr - 1.0, zr);
-        double d001 = gradDot(this.p(xy00 + z + 1), xr, yr, zr - 1.0);
-        double d101 = gradDot(this.p(xy10 + z + 1), xr - 1.0, yr, zr - 1.0);
-        double d011 = gradDot(this.p(xy01 + z + 1), xr, yr - 1.0, zr - 1.0);
-        double d111 = gradDot(this.p(xy11 + z + 1), xr - 1.0, yr - 1.0, zr - 1.0);
-        double xAlpha = Mth.smoothstep(xr);
-        double yAlpha = Mth.smoothstep(yrOriginal);
-        double zAlpha = Mth.smoothstep(zr);
-        return Mth.lerp3(xAlpha, yAlpha, zAlpha, d000, d100, d010, d110, d001, d101, d011, d111);
+        // DivineMC start - C2ME: optimize noise generation
+        final int var0 = x & 0xFF;
+        final int var1 = (x + 1) & 0xFF;
+        final int var2 = this.p[var0] & 0xFF;
+        final int var3 = this.p[var1] & 0xFF;
+        final int var4 = (var2 + y) & 0xFF;
+        final int var5 = (var3 + y) & 0xFF;
+        final int var6 = (var2 + y + 1) & 0xFF;
+        final int var7 = (var3 + y + 1) & 0xFF;
+        final int var8 = this.p[var4] & 0xFF;
+        final int var9 = this.p[var5] & 0xFF;
+        final int var10 = this.p[var6] & 0xFF;
+        final int var11 = this.p[var7] & 0xFF;
+
+        final int var12 = (var8 + z) & 0xFF;
+        final int var13 = (var9 + z) & 0xFF;
+        final int var14 = (var10 + z) & 0xFF;
+        final int var15 = (var11 + z) & 0xFF;
+        final int var16 = (var8 + z + 1) & 0xFF;
+        final int var17 = (var9 + z + 1) & 0xFF;
+        final int var18 = (var10 + z + 1) & 0xFF;
+        final int var19 = (var11 + z + 1) & 0xFF;
+        final int var20 = (this.p[var12] & 15) << 2;
+        final int var21 = (this.p[var13] & 15) << 2;
+        final int var22 = (this.p[var14] & 15) << 2;
+        final int var23 = (this.p[var15] & 15) << 2;
+        final int var24 = (this.p[var16] & 15) << 2;
+        final int var25 = (this.p[var17] & 15) << 2;
+        final int var26 = (this.p[var18] & 15) << 2;
+        final int var27 = (this.p[var19] & 15) << 2;
+        final double var60 = xr - 1.0;
+        final double var61 = yr - 1.0;
+        final double var62 = zr - 1.0;
+        final double var87 = FLAT_SIMPLEX_GRAD[(var20) | 0] * xr + FLAT_SIMPLEX_GRAD[(var20) | 1] * yr + FLAT_SIMPLEX_GRAD[(var20) | 2] * zr;
+        final double var88 = FLAT_SIMPLEX_GRAD[(var21) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var21) | 1] * yr + FLAT_SIMPLEX_GRAD[(var21) | 2] * zr;
+        final double var89 = FLAT_SIMPLEX_GRAD[(var22) | 0] * xr + FLAT_SIMPLEX_GRAD[(var22) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var22) | 2] * zr;
+        final double var90 = FLAT_SIMPLEX_GRAD[(var23) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var23) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var23) | 2] * zr;
+        final double var91 = FLAT_SIMPLEX_GRAD[(var24) | 0] * xr + FLAT_SIMPLEX_GRAD[(var24) | 1] * yr + FLAT_SIMPLEX_GRAD[(var24) | 2] * var62;
+        final double var92 = FLAT_SIMPLEX_GRAD[(var25) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var25) | 1] * yr + FLAT_SIMPLEX_GRAD[(var25) | 2] * var62;
+        final double var93 = FLAT_SIMPLEX_GRAD[(var26) | 0] * xr + FLAT_SIMPLEX_GRAD[(var26) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var26) | 2] * var62;
+        final double var94 = FLAT_SIMPLEX_GRAD[(var27) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var27) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var27) | 2] * var62;
+
+        final double var95 = xr * 6.0 - 15.0;
+        final double var96 = yrOriginal * 6.0 - 15.0;
+        final double var97 = zr * 6.0 - 15.0;
+        final double var98 = xr * var95 + 10.0;
+        final double var99 = yrOriginal * var96 + 10.0;
+        final double var100 = zr * var97 + 10.0;
+        final double var101 = xr * xr * xr * var98;
+        final double var102 = yrOriginal * yrOriginal * yrOriginal * var99;
+        final double var103 = zr * zr * zr * var100;
+
+        final double var113 = var87 + var101 * (var88 - var87);
+        final double var114 = var93 + var101 * (var94 - var93);
+        final double var115 = var91 + var101 * (var92 - var91);
+        final double var116 = var89 + var101 * (var90 - var89);
+        final double var117 = var114 - var115;
+        final double var118 = var102 * (var116 - var113);
+        final double var119 = var102 * var117;
+        final double var120 = var113 + var118;
+        final double var121 = var115 + var119;
+        return var120 + (var103 * (var121 - var120));
+        // DivineMC end - C2ME: optimize noise generation
     }
 
     private double sampleWithDerivative(final int x, final int y, final int z, final double xr, final double yr, final double zr, final double[] derivativeOut) {
diff --git a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
index 8ca042d2d16c3434154ba641a467c4e14c88328c..36dbb7522462910f04c643ce4626490be1f0eccd 100644
--- a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
@@ -26,6 +26,7 @@ public class PerlinNoise {
     public final double lowestFreqValueFactor;
     public final double lowestFreqInputFactor;
     private final double maxValue;
+    private final double [] amplitudesArray; // DivineMC - C2ME: optimize noise generation
 
     @Deprecated
     public static PerlinNoise createLegacyForBlendedNoise(final RandomSource random, final IntStream octaves) {
@@ -127,6 +128,7 @@ public class PerlinNoise {
         this.lowestFreqInputFactor = Math.pow(2.0, -zeroOctaveIndex);
         this.lowestFreqValueFactor = Math.pow(2.0, octaves - 1) / (Math.pow(2.0, octaves) - 1.0);
         this.maxValue = this.edgeValue(2.0);
+        this.amplitudesArray = this.amplitudes.toDoubleArray(); // DivineMC - C2ME: optimize noise generation
     }
 
     protected double maxValue() {
@@ -138,7 +140,29 @@ public class PerlinNoise {
     }
 
     public double getValue(final double x, final double y, final double z) {
-        return this.getValue(x, y, z, 0.0, 0.0);
+        // DivineMC start - C2ME: optimize noise generation
+        double d = 0.0;
+        double e = this.lowestFreqInputFactor;
+        double f = this.lowestFreqValueFactor;
+
+        final int length = this.noiseLevels.length;
+
+        for (int i = 0; i < length; i++) {
+            ImprovedNoise perlinNoiseSampler = this.noiseLevels[i];
+            if (perlinNoiseSampler != null) {
+                @SuppressWarnings("deprecation")
+                double g = perlinNoiseSampler.noise(
+                    wrap(x * e), wrap(y * e), wrap(z * e), 0.0, 0.0
+                );
+                d += this.amplitudesArray[i] * g * f;
+            }
+
+            e *= 2.0;
+            f /= 2.0;
+        }
+
+        return d;
+        // DivineMC end - C2ME: optimize noise generation
     }
 
     @Deprecated
@@ -186,7 +210,7 @@ public class PerlinNoise {
     }
 
     public static double wrap(final double x) {
-        return x - Mth.lfloor(x / 3.3554432E7 + 0.5) * 3.3554432E7;
+        return x - Math.floor(x / 3.3554432E7 + 0.5) * 3.3554432E7; // DivineMC - C2ME: optimize noise generation
     }
 
     protected int firstOctave() {
